Attempt concatenation when cubes have overlapping data

valeriupredoi opened the pull request:
Better concatenation for **two** cubes that have time overlapping data and identical metadata - solves https://github.com/ESMValGroup/ESMValTool/issues/1329 and the olde issue with overlapping CMIP data (ie 1850-2000 and 1850-2300 both files in the same repo)

BTW whoever implemented the `exp: list` functionality **should have tested* properly or added info to documentation, at the moment it's completely useless :beer:

**Tasks**

-   [x] [Create an issue](https://github.com/ESMValGroup/ESMValCore/issues) to discuss what you are going to do, if you haven't done so already (and add the link at the bottom)
-   [x] This pull request has a descriptive title that can be used in a changelog
-   [x] Add unit tests
-   [x] Public functions should have a numpy-style docstring so they appear properly in the [API documentation](https://esmvaltool.readthedocs.io/projects/esmvalcore/en/latest/api/esmvalcore.html). For all other functions a one line docstring is sufficient.
-   [x] If writing a new/modified preprocessor function, please update the [documentation](https://esmvaltool.readthedocs.io/projects/esmvalcore/en/latest/esmvalcore/preprocessor.html)
-   [x] Circle/CI tests pass. Status can be seen below your pull request. If the tests are failing, click the link to find out why.
-   [x] Codacy code quality checks pass. Status can be seen below your pull request. If there is an error, click the link to find out why. If you suspect Codacy may be wrong, please ask by commenting.

If you need help with any of the tasks above, please do not hesitate to ask by commenting in the issue or pull request.

* * *

Closes #349 and possibly (untested) #54

jvegasbsc commented:
I was the one that added the documentation, I will do my penitence adding tests for this feature

valeriupredoi commented:
hahaha no worries @jvegasbsc am almost finished writing tests :beer:

jvegasbsc commented:
It is already done

jvegasbsc commented:
I have to add #259 to make the tests pass, as I was using 0:00 hours in the time

valeriupredoi commented:
sweeet! I was just about to commit, but not no more, cheers Javi :beer:

valeriupredoi commented:
@jvegasbsc I fixed your fix, man :grin:
The way was implemented was not working unfortunately, now it's all very schoolboy explicit for all possible cases. I removed an older test since we can do concatenation of overlapping cubes now; all your tests work; I also tested with real life data :beer:

mathause commented on esmvalcore/preprocessor/_io.py:
```diff
@@ -274,3 +289,55 @@ def _write_ncl_metadata(output_dir, metadata):
+def _concatenate_two_overlapping_cubes(cubes):
+    """Concatenate time-overlapping cubes (two cubes only)."""
+    # get time end points
+    time_1 = cubes[0].coord('time')
```
maybe rename to `time_0`, `time_1`, etc. so it is in line with the indices.

valeriupredoi replied:
meh, it's cube1, cube2 etc :grin:

valeriupredoi commented:
@mathause cheers muchly for testing!! good to hear you are getting the (mostly) desired output, yeah, I'll have a look at the changes next week, off to pub now :beer:

mathause commented:
Could not get to work right now, have to come back on monday...

valeriupredoi commented:
@schlunma remined me of #54 that this PR should solve as well :beer:

mattiarighi commented:
Please consider merging #210 first, since it changes a bit the way input files are handled.

valeriupredoi commented:
> Please consider merging #210 first, since it changes a bit the way input files are handled.

@mattiarighi it shouldn't matter, this one here only deals with the concatenation in `_io.py` after files have been found

bouweandela commented:
Does this close #54?

jvegasbsc commented:
> Does this close #54?
Not really, as we are only working on the case of two overlapping cubes, mostly for concatenation of experiments.

Some of the specific problems listed (in particular, cases like `tasmax_day_IPSL-CM5A-LR_historical_r1i1p1_19500101-19991231.nc`
`tasmax_day_IPSL-CM5A-LR_historical_r1i1p1_19500101-20051231.nc`) will be solved but mos of them no

valeriupredoi commented:
note however that the concatenation function will work with any case of overlapping data, so it's not just for different experiments :beer:

valeriupredoi commented:
#210 merged, let's get this baby in @bouweandela @mattiarighi @schlunma :beer:

bouweandela commented on esmvalcore/preprocessor/_io.py:
```diff
@@ -90,6 +91,16 @@ def concatenate(cubes):
+            pass
+        else:
+            concatenated = _concatenate_two_overlapping_cubes(concatenated)
+            if len(concatenated) == 1:
```
these two lines are duplicated, if you move the `if len(concatenated) == 2` block before the `if len(concatenated) == 1` statement above, they are not needed.

valeriupredoi replied:
no, there are two possible cases stemming from two different concatenation processes: the first if check - `if len(concatenated) == 1: return concatenated[0]` comes from simple concatenation and if that returns false then we check on the length to be exactly 2 - `if len(concatenated) == 2: try: except: else: if len(concatenated) == 1: return concatenated[0]` note that the seemingly repetive check on the lenght = 1 comes from the `_concatenate_two_overlapping_cubes`

bouweandela replied:
There is really no need to check `if len(concatenated) == 1` two times, can you have another look please?

bouweandela commented on tests/integration/preprocessor/_io/test_concatenate.py:
```diff
@@ -14,33 +14,41 @@ class TestConcatenate(unittest.TestCase):
-        self.raw_cubes.append(self.raw_cubes[0].copy())
-        with self.assertRaises(ValueError):
-            _io.concatenate(self.raw_cubes)
+    def test_concatenate_with_overlap(self):
```
You still need to test for failing if there is more than 2 cubes

valeriupredoi replied:
that is taken care by the older test

bouweandela replied:
which you removed in this pull request..

valeriupredoi replied:
dang! did I? :man_facepalming:

valeriupredoi commented:
@bouweandela cheers for the review, dude. It seems to me that you are not quite fully understanding the functionality of this implementation - yes, we need to test twice for the length of cubeList since once is from the actual standard concatenation (and the smart concatenation can only be performed if that returns 2 cubes) and the second time is for the actual result of the smart concatenation - if it's one cube we have bingo! if not we let the function fail and throw exception - this comment also takes care of your second comment :beer:

valeriupredoi commented:
OK documentation needs to be addressed indeed :beer:

valeriupredoi commented:
Mr @bouweandela you are to find a myriad of tests now :grin: Where do you think we should put the Concatenation info in the doc? I didn't seem to create and deciated section called Concatenation when I rehauled the preprocessor documentation :beer:

JaroCamphuijsen commented on esmvalcore/preprocessor/_io.py:
```diff
@@ -269,3 +280,63 @@ def _write_ncl_metadata(output_dir, metadata):
+    if cubes[0].coord('time').points[0] <= cubes[1].coord('time').points[0]:
+        cubes = [cubes[0], cubes[1]]
+    else:
+        cubes = [cubes[1], cubes[0]]
```
@valeriupredoi This case is not covered in the tests yet.

JaroCamphuijsen commented on esmvalcore/preprocessor/_io.py:
```diff
@@ -269,3 +280,63 @@ def _write_ncl_metadata(output_dir, metadata):
+        if data_end_1 < data_end_2:
+            cubes = [cubes[1]]
+        else:
+            cubes = [cubes[0]]
```
@valeriupredoi This case is not covered in the tests yet.

valeriupredoi replied:
good call, mate! Now it has full coverage :beer:

bouweandela commented on doc/esmvalcore/datafinder.rst:
```diff
@@ -257,3 +257,22 @@ Since observational data are organized in Tiers depending on their level of
+frequent with data from models overlapping in time, and is accounted for by a function that performs a
+flexible concatenation between two cubes, depending on the particular setup:
+
+* cubes overlap time: resulting cube is made up of the overlapping data plus left and
```
Can you explain here how you choose which data to use? This matters in case the overlapping data is not the same.

valeriupredoi replied:
just did, good call, man!

valeriupredoi commented:
really need an approve here folks @mattiarighi @ledm @schlunma and anyone else pls

bouweandela commented on esmvalcore/preprocessor/_io.py:
```diff
@@ -269,3 +281,63 @@ def _write_ncl_metadata(output_dir, metadata):
+             if t in time_2.points),
+            None
+        )
+        # case 0: no overlap (new iris implementaion does allow
```
This seems to conflict with `# case 2.3: there is no overlap: return original cubes` defined on line 341. Also, I think this check is not needed here, because a discontinuous time coordinate is something that should be found by the cmor check.

valeriupredoi replied:
Yeah that line 341 is not actually a case per-se (need to remove the comment) - it's safer to raise the error if no overlap than to allow the concatenation to happen and then have it found by cmor.check when looking for time discontinuities

bouweandela replied:
I'm not sure, because in the usual case no overlap is expected. E.g. if concatenation fails because there are problems related to coordinates other than the time coordinate, you will get this error, which is really confusing..

valeriupredoi replied:
Ermm, that is true, but then again, having no overlap in time is still a valid issue with the cubes that must be reported...

valeriupredoi replied:
Oh actually, yeah, you're right! in a standard case there is no overlap and the lack of it will be reported - we need to remove that error, maybe make it a logger debug or warning?

bouweandela replied:
I think it's enough to simply remove the check and exception here

bouweandela commented on tests/integration/preprocessor/_io/test_concatenate.py:
```diff
@@ -14,34 +16,174 @@ class TestConcatenate(unittest.TestCase):
+        with self.assertRaises(TypeError):
             _io.concatenate(self.raw_cubes)
 
+    def test_fail_on_units_concatenate_with_overlap(self):
```
Why should this fail? Can't you can just convert all the time coordinates to the same units?

valeriupredoi replied:
Yes, but this is a catch so that in case there are situations that are not picked up by cmor.fix

bouweandela commented on tests/integration/preprocessor/_io/test_concatenate.py:
```diff
@@ -14,34 +16,174 @@ class TestConcatenate(unittest.TestCase):
+        concatenated = _io.concatenate(cubes_ordered)
+        self.assertTrue(np.allclose(
+            concatenated.coord('time').points,
+            np.array([1., 2., 5., 7., 100.])
```
It looks like the time point at 1.5 as defined on line 112 is simply discarded, I'm not sure this is the kind of behaviour that users would find intuitive

valeriupredoi replied:
Yeh well, it's part of the plan :grin:

bouweandela commented on tests/integration/preprocessor/_io/test_concatenate.py:
```diff
@@ -14,34 +16,174 @@ class TestConcatenate(unittest.TestCase):
+        concatenated = _io.concatenate(self.raw_cubes)
+        self.assertTrue(np.allclose(
+            concatenated.coord('time').points,
+            np.array([1., 2., 3., 4., 5., 6., 1000., 7000.])
```
Where did the time point 200 as defined on line 57 go?

valeriupredoi replied:
Thing is, the overlap can not be one single border point, so the [3, 200] cube is not counted in

bouweandela replied:
At the very least, I think it would be good to add some warning log messages telling the user that some data is ignored. 

If there is overlapping data, I think it would be nice to have a debug log message telling the user what data from what cube you're using.

valeriupredoi replied:
Yeah that'd be very sensible, but it's a bit of a headache. I'll see what I can do after Xmas :grin: I still think we should get this in sopner than later because it's quite needed, maybe we can open an issue and a new PR for these logger mssges after merge?

bouweandela replied:
I would prefer to have it working properly before merging. Silently ignoring data makes the tool really hard to understand.

valeriupredoi commented:
OK @bouweandela debug messages added - now it should be very clear what data is being used and how :beer: Can we go ahead pls :medal_military:

bouweandela commented on esmvalcore/preprocessor/_io.py:
```diff
@@ -269,3 +281,88 @@ def _write_ncl_metadata(output_dir, metadata):
+    # case 1: both cubes start at the same time -> return longer cube
+    if data_start_1 == data_start_2:
+        if data_end_1 < data_end_2:
+            logger.debug(f"Both cubes start at the same time but cube %s "
```
```suggestion
            logger.debug("Both cubes start at the same time but cube %s "
```
no need to make an f-string if you're using it as a normal string

bouweandela commented on esmvalcore/preprocessor/_io.py:
```diff
@@ -269,3 +281,88 @@ def _write_ncl_metadata(output_dir, metadata):
+        # case 0: no overlap (new iris implementaion does allow
+        # concatenation of cubes with no overlap)
+        if not start_overlap:
+            raise ValueError(f"Attempting to concatenate cubes that are"
```
```suggestion
            logger.debug("Unable to concatenate non-overlapping cubes\n%s\nand\n%s"
            return cubes
```
because of this https://github.com/ESMValGroup/ESMValCore/pull/280/files#r360837753

bouweandela commented:
@valeriupredoi I made some minor changes, can you check if you're still happy with the result?

valeriupredoi commented on esmvalcore/preprocessor/_io.py:
```diff
@@ -305,52 +307,56 @@ def _concatenate_overlapping_cubes(cubes):
-                         f"ends before %s", cubes[0], cubes[1])
-            logger.debug(f"Cube %s contains all needed data "
-                         f"so using it fully", cubes[1])
+        if data_end_1 <= data_end_2:
```
good catch!

valeriupredoi commented:
@mattiarighi you wanna test pls mate? :beer:
